File: app-android/vela-wallet/app/src/main/java/app/getvela/wallet/feature/signing/core/ClearWire.kt
package app.getvela.wallet.feature.signing.core

import app.getvela.wallet.core.format.DateFormatKey
import app.getvela.wallet.core.format.Formats
import app.getvela.wallet.core.format.NumberFormatKey
import app.getvela.wallet.core.format.TimeFormatKey
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/*
 * The `clear_signing` machine's wire (spec 044) — what a request DOES and
 * how dangerous it is. `chain_id` is `u32`, `usd_value` `f64`, a timer's
 * `ms`/`token` `u32`, `tz_offset_minutes` `i32`.
 */

@Serializable
enum class ClearFieldRole {
    @SerialName("send_amount") SendAmount,

    @SerialName("receive_amount") ReceiveAmount,

    @SerialName("recipient") Recipient,

    @SerialName("spender") Spender,

    @SerialName("generic") Generic,
}

@Serializable
enum class ClearSignType {
    @SerialName("transaction") Transaction,

    @SerialName("signature") Signature,
}

@Serializable
enum class ClearSignMethod {
    @SerialName("personal_sign") PersonalSign,

    @SerialName("eth_sign") EthSign,
}

@Serializable
enum class ClearRisk {
    @SerialName("safe") Safe,

    @SerialName("normal") Normal,

    @SerialName("caution") Caution,

    @SerialName("danger") Danger,
}

@Serializable
enum class ClearDangerClass {
    @SerialName("plain") Plain,

    @SerialName("siwe_ok") SiweOk,

    @SerialName("siwe_phish") SiwePhish,

    @SerialName("opaque_hash") OpaqueHash,

    @SerialName("eth_sign") EthSign,
}

@Serializable
enum class ClearSurface {
    @SerialName("none") None,

    @SerialName("loading") Loading,

    @SerialName("clear_sign") ClearSign,

    @SerialName("eth_sign") EthSign,

    @SerialName("message_sign") MessageSign,

    @SerialName("blind_typed_data") BlindTypedData,

    @SerialName("blind_transaction") BlindTransaction,
}

@Serializable
enum class ClearProbe {
    @SerialName("supports_erc721") SupportsErc721,

    @SerialName("supports_erc1155") SupportsErc1155,

    @SerialName("decimals") Decimals,

    @SerialName("symbol") Symbol,
}

@Serializable
enum class ClearSiweBinding {
    @SerialName("ok") Ok,

    @SerialName("mismatch") Mismatch,

    @SerialName("unknown") Unknown,
}

@Serializable
enum class ClearDateFormat {
    @SerialName("ymd_slash") YmdSlash,

    @SerialName("iso") Iso,

    @SerialName("dmy_slash") DmySlash,

    @SerialName("dmy_dot") DmyDot,

    @SerialName("mdy_slash") MdySlash,
}

@Serializable
enum class ClearNumberFormat {
    @SerialName("comma_dot") CommaDot,

    @SerialName("dot_comma") DotComma,

    @SerialName("space_comma") SpaceComma,

    @SerialName("indian") Indian,
}

@Serializable
enum class ClearTimeFormat {
    @SerialName("h24") H24,

    @SerialName("h12") H12,
}

@Serializable
data class ClearLocale(
    val number_format: ClearNumberFormat = ClearNumberFormat.CommaDot,
    val date_format: ClearDateFormat = ClearDateFormat.Iso,
    val time_format: ClearTimeFormat = ClearTimeFormat.H24,
    val tz_offset_minutes: Int = 0,
) {
    companion object {
        /**
         * Spec 049: the person's resolved presets and the phone's UTC offset
         * now — the web's `toClearLocale(resolvedFormatKeys())` shape. The
         * constant that stood in before printed a dApp's amounts in a format
         * the person had not chosen.
         */
        fun fromFormats(formats: Formats, nowMs: Long = System.currentTimeMillis()): ClearLocale = ClearLocale(
            number_format = when (formats.resolvedNumber()) {
                NumberFormatKey.DotComma -> ClearNumberFormat.DotComma
                NumberFormatKey.SpaceComma -> ClearNumberFormat.SpaceComma
                NumberFormatKey.Indian -> ClearNumberFormat.Indian
                NumberFormatKey.CommaDot, NumberFormatKey.Auto -> ClearNumberFormat.CommaDot
            },
            date_format = when (formats.resolvedDate()) {
                DateFormatKey.YmdSlash -> ClearDateFormat.YmdSlash
                DateFormatKey.MdySlash -> ClearDateFormat.MdySlash
                DateFormatKey.DmySlash -> ClearDateFormat.DmySlash
                DateFormatKey.DmyDot -> ClearDateFormat.DmyDot
                DateFormatKey.Iso, DateFormatKey.Auto -> ClearDateFormat.Iso
            },
            time_format = if (formats.resolvedTime() == TimeFormatKey.H12) ClearTimeFormat.H12 else ClearTimeFormat.H24,
            // Minutes to ADD to UTC (the web negates `getTimezoneOffset()`).
            tz_offset_minutes = java.util.TimeZone.getDefault().getOffset(nowMs) / 60_000,
        )
    }
}

@Serializable
sealed class ClearConfirm {
    @Serializable
    @SerialName("sign")
    data object Sign : ClearConfirm()

    @Serializable
    @SerialName("confirm")
    data object Confirm : ClearConfirm()

    @Serializable
    @SerialName("confirm_intent")
    data class ConfirmIntent(val intent: String) : ClearConfirm()
}

@Serializable
data class ClearSignField(
    val label: String,
    val value: String,
    val format: String = "",
    val token_address: String? = null,
    val warning: Boolean = false,
    val unverified: Boolean = false,
    val role: ClearFieldRole = ClearFieldRole.Generic,
    val detail: Boolean = false,
    val expired: Boolean = false,
    val address: String? = null,
    val usd_value: Double? = null,
)

@Serializable
data class ClearSignResult(
    val intent: String,
    val contract_name: String? = null,
    val owner: String? = null,
    val fields: List<ClearSignField> = emptyList(),
    val risk: ClearRisk = ClearRisk.Normal,
    val contract_address: String? = null,
    val verified: Boolean = false,
    val sign_type: ClearSignType = ClearSignType.Transaction,
    val partial: Boolean = false,
    val best_effort: Boolean = false,
    val to_own_token: Boolean = false,
)

@Serializable
data class ClearBlindField(val key: String, val value: String)

@Serializable
data class ClearBlindTyped(
    val primary_type: String? = null,
    val has_domain: Boolean = false,
    val domain_name: String? = null,
    val verifying_contract: String? = null,
    val fields: List<ClearBlindField> = emptyList(),
)

@Serializable
data class ClearSiweFields(
    val domain: String,
    val domain_host: String? = null,
    val address: String? = null,
    val statement: String? = null,
    val uri: String? = null,
    val chain_id: Int? = null,
    val nonce: String? = null,
)

@Serializable
data class ClearMessageView(
    val payload: String,
    val is_hex: Boolean = false,
    val decoded_text: String? = null,
    val binary_preview: String? = null,
    val non_printable: Boolean = false,
    val siwe: ClearSiweFields? = null,
    val binding: ClearSiweBinding? = null,
    val danger_class: ClearDangerClass = ClearDangerClass.Plain,
)

@Serializable
data class ClearSigningView(
    val resolving: Boolean = false,
    val resolved: Boolean = false,
    val result: ClearSignResult? = null,
    val message: ClearMessageView? = null,
    val surface: ClearSurface = ClearSurface.None,
    val confirm: ClearConfirm = ClearConfirm.Confirm,
    val blind_typed: ClearBlindTyped? = null,
    val danger_haptic: Boolean = false,
)

@Serializable
sealed class ClearOperation {
    @Serializable
    @SerialName("http_get")
    data class HttpGet(val path: String) : ClearOperation()

    @Serializable
    @SerialName("rpc_eth_call")
    data class RpcEthCall(val chain_id: Int, val to: String, val data: String, val probe: ClearProbe) : ClearOperation()

    @Serializable
    @SerialName("selector_db_lookup")
    data class SelectorDbLookup(val selector: String) : ClearOperation()

    @Serializable
    @SerialName("timer")
    data class Timer(val ms: Int, val token: Int) : ClearOperation()

    @Serializable
    @SerialName("now")
    data object Now : ClearOperation()
}

@Serializable
sealed class ClearShellResult {
    @Serializable
    @SerialName("descriptor_fetched")
    data class DescriptorFetched(val path: String, val json: String? = null) : ClearShellResult()

    @Serializable
    @SerialName("rpc_answer")
    data class RpcAnswer(
        val probe: ClearProbe,
        val chain_id: Int,
        val to: String,
        val result: String? = null,
        val rpc_error: Boolean = false,
    ) : ClearShellResult()

    @Serializable
    @SerialName("selector_candidates")
    data class SelectorCandidates(val sigs: List<String> = emptyList()) : ClearShellResult()

    @Serializable
    @SerialName("timed_out")
    data class TimedOut(val token: Int) : ClearShellResult()

    @Serializable
    @SerialName("clock")
    data class Clock(val now_ms: Double) : ClearShellResult()
}

@Serializable
sealed class ClearSigningEvent {
    @Serializable
    @SerialName("resolve_transaction")
    data class ResolveTransaction(
        val to: String? = null,
        val data: String? = null,
        val value: String? = null,
        val chain_id: Int,
        val locale: ClearLocale = ClearLocale(),
    ) : ClearSigningEvent()

    @Serializable
    @SerialName("resolve_typed_data")
    data class ResolveTypedData(val typed_data_json: String, val chain_id: Int, val locale: ClearLocale = ClearLocale()) : ClearSigningEvent()

    @Serializable
    @SerialName("message_presented")
    data class MessagePresented(val method: ClearSignMethod, val params: List<String>, val request_origin: String? = null) : ClearSigningEvent()

    @Serializable
    @SerialName("cleared")
    data object Cleared : ClearSigningEvent()
}

File: app-ios/VelaWallet/VelaWallet/Features/Signing/Core/ClearWire.swift
//
//  ClearWire.swift
//  VelaWallet
//
//  The `clear_signing` machine's view model, in Swift.
//
//  What a transaction DOES, in words, before it says what it is in hex. Five
//  thousand lines of core decide that — the descriptor fetch, the 4-byte
//  fallback, the chain probes, the six degradation rungs — and none of it is
//  re-derived here.
//
//  ## `surface` is the dispatch, not a hint
//
//  The sheet's full order is: typed permit → editable approval → LOADING →
//  CLEAR SIGN → batch → ETH_SIGN → MESSAGE → BLIND TYPED → BLIND TX. The two
//  approval surfaces and the batch belong to `approval_guard`; everything in
//  capitals is decided here. The sheet interleaves exactly two verdicts and
//  decides nothing itself.
//
//  `loading` holds the sheet on purpose: a blind view must never flash before
//  the clear one, because somebody reading fast would see the scary version of
//  a transaction that turns out to be fine, or worse, the calm version of one
//  that is not.
//
//  Views are `Decodable` through `CoreJSON.decoder`; operations and results
//  stay dictionaries.
//

import Foundation

/// Risk level for visual treatment.
enum ClearRisk: String, Decodable {
    case safe, normal, caution, danger
}

/// Layout role hint — which slot a field belongs in.
enum ClearFieldRole: String, Decodable {
    case sendAmount = "send_amount"
    case receiveAmount = "receive_amount"
    case recipient
    case spender
    case generic
}

enum ClearSignType: String, Decodable {
    case transaction, signature
}

/// One resolved display field.
struct ClearSignFieldWire: Decodable, Equatable {
    let label: String
    let value: String
    let format: String
    /// A normalised, validated token address, for the logo lookup.
    let tokenAddress: String?
    /// A high-risk field — an unlimited approval, say. Renders as danger.
    let warning: Bool
    /// An amount shown with decimals that were never verified on-chain.
    /// Caution, not danger: it may well be right, and it may be off by
    /// several orders of magnitude.
    let unverified: Bool
    let role: ClearFieldRole
    /// Collapsed under the advanced disclosure.
    let detail: Bool
    /// A deadline already in the past.
    let expired: Bool
    /// The full lowercased address, for address-name fields.
    let address: String?
    let usdValue: Double?
}

/// The resolved result, ready to draw.
struct ClearSignResultWire: Decodable, Equatable {
    /// The canonical English key — the shell localises it. Printing this
    /// verbatim is how Android shipped a confirm button reading "确认send".
    let intent: String
    let contractName: String?
    let owner: String?
    let fields: [ClearSignFieldWire]
    let risk: ClearRisk
    let contractAddress: String?
    let verified: Bool
    let signType: ClearSignType
    /// The descriptor declared more fields than resolved. Say "incomplete"
    /// loudly rather than showing a partial list as if it were whole.
    let partial: Bool
    /// Recovered through the 4-byte database and decoded generically. Best
    /// effort, and it must read that way.
    let bestEffort: Bool
    /// A recipient of this call **is** the contract being called: the token
    /// is being sent to its own contract, which burns it irreversibly.
    let toOwnToken: Bool
}

enum ClearSignMethod: String, Decodable {
    case personalSign = "personal_sign"
    /// `eth_sign` signs an opaque hash. It gets the hard-warning surface,
    /// never the calm message view.
    case ethSign = "eth_sign"
}

/// Parsed EIP-4361 sign-in fields.
struct ClearSiweFieldsWire: Decodable, Equatable {
    let domain: String
    /// The lowercased host the binding was actually **compared on**, or `nil`
    /// when the authority is unparseable. The domain row renders this, so the
    /// string on screen is the string that was adjudicated — showing a
    /// prettier one than the check ran against is how a lookalike slips past.
    let domainHost: String?
    let address: String?
    let statement: String?
    let uri: String?
    let chainId: Int?
    let nonce: String?
}

enum ClearSiweBinding: String, Decodable {
    case ok, mismatch, unknown
}

/// Which danger treatment the message surface takes.
///
/// `siweOk` covers both a matching binding and an unknown one — the calm
/// sign-in layout. The verified badge renders only on a match, and only a
/// **mismatch** escalates. An unknown origin never asserts a match, and is
/// not evidence of phishing either.
enum ClearDangerClass: String, Decodable {
    case plain
    case siweOk = "siwe_ok"
    case siwePhish = "siwe_phish"
    case opaqueHash = "opaque_hash"
    case ethSign = "eth_sign"
}

/// Everything the message view needs, computed once.
struct ClearMessageViewWire: Decodable, Equatable {
    /// The param this method actually signs, **already chosen**: `params[0]`
    /// for `personal_sign`, `params[1]` for `eth_sign`, falling back to
    /// `params[0]` for a malformed single-param `eth_sign`. The shell does not
    /// pick.
    let payload: String
    let isHex: Bool
    /// Hex decoded as UTF-8, or the verbatim non-hex payload.
    let decodedText: String?
    /// A short preview for genuinely binary payloads.
    let binaryPreview: String?
    let nonPrintable: Bool
    let siwe: ClearSiweFieldsWire?
    let binding: ClearSiweBinding?
    let dangerClass: ClearDangerClass
}

/// One raw entry of an undecodable EIP-712 payload, already rendered to the
/// single line the sheet shows.
struct ClearBlindFieldWire: Decodable, Equatable {
    let key: String
    let value: String
}

/// The blind typed-data projection.
///
/// Deliberately **not** reinterpreted: no decimals, no timestamp guessing. The
/// descriptor is unknown, so an honest raw value beats a confident wrong one.
struct ClearBlindTypedWire: Decodable, Equatable {
    let primaryType: String?
    let hasDomain: Bool
    let domainName: String?
    let verifyingContract: String?
    /// The first five message entries, in payload order.
    let fields: [ClearBlindFieldWire]
}

enum ClearSurface: String, Decodable {
    /// Nothing presented, or a method this machine does not adjudicate.
    case none
    /// A descriptor is resolving. **Holds the sheet.**
    case loading
    case clearSign = "clear_sign"
    case ethSign = "eth_sign"
    case messageSign = "message_sign"
    case blindTypedData = "blind_typed_data"
    case blindTransaction = "blind_transaction"
}

/// The confirm button's **semantics**. The words stay in the shell, and so
/// does the measurement of whether a localised intent fits — that is a
/// property of the translated string, not of the request.
enum ClearConfirmWire: Decodable, Equatable {
    /// A pure signature.
    case sign
    /// Neutral. **Never "approve"**: that verb belongs only to an actual token
    /// approval, which is `approval_guard`'s surface.
    case confirm
    /// "Confirm {intent}", where `intent` is the canonical English key for
    /// the shell to localise — never printed raw.
    case confirmIntent(String)

    private enum Keys: String, CodingKey { case type, intent }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: Keys.self)
        switch try container.decode(String.self, forKey: .type) {
        case "sign": self = .sign
        case "confirm": self = .confirm
        case "confirm_intent":
            self = .confirmIntent(try container.decode(String.self, forKey: .intent))
        case let other:
            throw DecodingError.dataCorruptedError(
                forKey: .type, in: container,
                debugDescription: "unknown clear confirm '\(other)'"
            )
        }
    }
}

struct ClearSigningViewWire: Decodable, Equatable {
    let resolving: Bool
    let resolved: Bool
    let result: ClearSignResultWire?
    let message: ClearMessageViewWire?
    let surface: ClearSurface
    let confirm: ClearConfirmWire
    /// Present for every typed-data request; reached only once resolution has
    /// concluded with no descriptor.
    let blindTyped: ClearBlindTypedWire?
    /// The sheet buzzes a warning on open. Computed **once**, here, so the
    /// haptic and the red banner can never disagree. The unbounded-approval
    /// half of the same buzz is `approval_guard`'s verdict; the sheet ORs the
    /// two and decides nothing.
    let dangerHaptic: Bool

    static let empty = ClearSigningViewWire(
        resolving: false, resolved: false, result: nil, message: nil,
        surface: .none, confirm: .confirm, blindTyped: nil, dangerHaptic: false
    )
}
